### PR TITLE
Development

### PR DESCRIPTION
**Pull request by amir62hossein**

### CODE REVIEW

File: controller/wallet_handler.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	return &WalletHandler{
+		svc : svc,
+	}
+}
+func AddWalletHRoutes(e *echo.Echo)  {
+	walletHandler := NewWalletHandler()
+	e.POST("/createWallet"  , walletHandler.GenerateWalletHandler)
+	e.PUT("/walletCharge" , walletHandler.ChargeWalletHandler)
+	e.GET("/walletAmount" , walletHandler.GetWalletAmountHandler)
+	e.GET("/walletCredit" , walletHandler.HasWalletCreditHandler)
```
Comment (cyneptic) [inline controller/wallet_handler.go]: 1. do not use camelCase in endpoint naming, either do lowercase or lowercase with hyphens ("createwallet", "create-wallet")
2. maybe do "wallet-create" intead of "create-wallet" to be consistent
3. "wallet-credit" is not a good name in relation to what it does, try something better maybe "wallet-can-afford" or "wallet-has-amount" or something similar.

File: controller/wallet_handler.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+func (w *WalletHandler) GenerateWalletHandler(c echo.Context) error {
+	paramID := c.QueryParam("id")
+	userID , err := uuid.Parse(paramID)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest , err.Error())
+	}
```
Comment (cyneptic) [inline controller/wallet_handler.go]: validation should go in validators.

File: controller/wallet_handler.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	amountStr := c.QueryParam("amount")
+
+	amount , err := strconv.Atoi(amountStr)
+	if err != nil {
+		// Handle the error if the conversion fails
+		return c.JSON(http.StatusBadRequest, "Invalid amount")
+	}
```
Comment (cyneptic) [inline controller/wallet_handler.go]: validation should go in validators

File: controller/wallet_handler.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	walletID := c.QueryParam("walletId")
+	walletUUId , err := uuid.Parse(walletID)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest , err.Error())
+	}
+	userID := c.QueryParam("userId")
+	userUUId , err := uuid.Parse(userID)
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, err.Error())
+	}
```
Comment (cyneptic) [inline controller/wallet_handler.go]: validation should go in validators.

File: controller/wallet_handler.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	if err != nil {
+		return c.JSON(http.StatusBadRequest , err.Error())
+	}
+	amountStr := c.QueryParam("amount")
+
+	amount , err := strconv.Atoi(amountStr)
+	if err != nil {
+		// Handle the error if the conversion fails
+		return c.JSON(http.StatusBadRequest, "Invalid amount")
+	}
```
Comment (cyneptic) [inline controller/wallet_handler.go]: validation should go in validators.

File: infrastructure/repository/wallet.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
+package repositories
+
+import (
+	"fmt"
+	"github.com/cyneptic/letsgo-smspanel/internal/core/entities"
+	"github.com/google/uuid"
+	"gorm.io/gorm"
+)
+
+func (r *PGRepository) GenerateWalletRepository(wallet entities.Wallet) error {
```
Comment (cyneptic) [inline infrastructure/repository/wallet.go]: Using repository in the function name makes one thing you are creating a wallet repository not a wallet entity within the repository
"GenerateWallet" or "CreateWallet" is better.

File: infrastructure/repository/wallet.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+	result := r.DB.Create(&wallet)
+	if result.Error != nil {
+		return result.Error
+	}
+	return nil
+}
+func (r *PGRepository) ChargeWalletRepository(walletID , userID uuid.UUID , amount int)  error {
```
Comment (cyneptic) [inline infrastructure/repository/wallet.go]: Using repository in the function name makes one thing you are charing a wallet repository not a wallet entity within the repository
"ChargeWallet" is better.

File: infrastructure/repository/wallet.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	// 3. Update the wallet's credit by adding the amount
+	wallet.Credit += amount
+	err = r.DB.Save(&wallet).Error
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+func (r *PGRepository) GetWalletAmountRepository(walletID  uuid.UUID) (*entities.Wallet, error) {
```
Comment (cyneptic) [inline infrastructure/repository/wallet.go]: if this function is returning a wallet it should be named "GetWallet" not "GetWalletAmount"

File: infrastructure/repository/wallet.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	err := r.DB.First(&wallet, "id = ?", walletID).Error
+	if err != nil {
+		if err == gorm.ErrRecordNotFound {
+			return nil , fmt.Errorf("wallet not found")
+		}
+		return nil , err
+	}
+	return &wallet, nil
+}
+func (r *PGRepository) HasEnoughCreditRepository(walletID uuid.UUID) (*entities.Wallet, error) {
```
Comment (cyneptic) [inline infrastructure/repository/wallet.go]: if this function is returning a wallet it should be named "GetWallet" not "HasEnoughCredit"

File: .idea/.gitignore
Comment (cyneptic) [inline .idea/.gitignore:1]: should this file be included?

File: .idea/vcs.xml
Comment (cyneptic) [inline .idea/vcs.xml:1]: should this file be included?

File: .idea/modules.xml
Comment (cyneptic) [inline .idea/modules.xml:1]: should this file be included?

File: .idea/letsgo_smspanel.iml
Comment (cyneptic) [inline .idea/letsgo_smspanel.iml:1]: should this file be included?

File: controller/contact_handlers.go
Comment (cyneptic) [inline controller/contact_handlers.go:1]: revert the changes to someone else's file there is no need to delete these spaces. specially if you are not the author of the code.

File: internal/core/service/wallet_service.go [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+	generatedWallet.CreatedAt = time.Now()
+	generatedWallet.Credit = 0
+	err := r.db.GenerateWalletRepository(generatedWallet)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+func (r *WalletService) ChargeWalletService(walletID, userID uuid.UUID , amount int) error  {
+
```
Comment (cyneptic) [inline internal/core/service/wallet_service.go:34]: should check if amount is a proper positive value, it could be negative

**Review by cyneptic — CHANGES_REQUESTED** [review]

**Comment by amir62hossein** [comment]
Meo Meo

**Review by cyneptic — APPROVED** [review]
lgtm

**Pull request merged**